oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,397 @@
+    }
+  }
+
+  private static String getHexEncodedSha256Hash(byte[] bytes) {
+    try {
+      MessageDigest digest = MessageDigest.getInstance("SHA-256");
+      return BaseEncoding.base16().lowerCase().encode(digest.digest(bytes));
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException("Failed to compute SHA-256 hash.", e);
+    }
+  }
+
+  static Builder newBuilder(
+      AwsSecurityCredentials awsSecurityCredentials, String httpMethod, String url, String region) {
+    return new Builder(awsSecurityCredentials, httpMethod, url, region);
+  }
+
+  static class Builder {
+
+    private AwsSecurityCredentials awsSecurityCredentials;
+    private String httpMethod;
+    private String url;
+    private String region;
+
+    @Nullable private String requestPayload;
+    @Nullable private Map<String, String> additionalHeaders;
+    @Nullable private AwsDates dates;
+
+    private Builder(
+        AwsSecurityCredentials awsSecurityCredentials,
+        String httpMethod,
+        String url,
+        String region) {
+      this.awsSecurityCredentials = awsSecurityCredentials;
+      this.httpMethod = httpMethod;
+      this.url = url;
+      this.region = region;
+    }
+
+    Builder setRequestPayload(String requestPayload) {
+      this.requestPayload = requestPayload;
+      return this;
+    }
+
+    Builder setAdditionalHeaders(Map<String, String> additionalHeaders) {
+      if (additionalHeaders.containsKey("date") && additionalHeaders.containsKey("x-amz-date")) {
+        throw new IllegalArgumentException("One of {date, x-amz-date} can be specified, not both.");
+      }
+      try {
+        if (additionalHeaders.containsKey("date")) {
+          this.dates = AwsDates.fromDateHeader(additionalHeaders.get("date"));
+        }
+        if (additionalHeaders.containsKey("x-amz-date")) {
+          this.dates = AwsDates.fromXAmzDate(additionalHeaders.get("x-amz-date"));
+        }
+      } catch (ParseException e) {
+        throw new IllegalArgumentException("The provided date header value is invalid.", e);
+      }
+
+      this.additionalHeaders = additionalHeaders;
+      return this;
+    }
+
+    AwsRequestSigner build() {
+      return new AwsRequestSigner(
+          awsSecurityCredentials,
+          httpMethod,
+          url,
+          region,
+          requestPayload,
+          additionalHeaders,
+          dates);
+    }
+  }
+
+  static final class AwsDates {
+
+    private static final String X_AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String CUSTOM_DATE_FORMAT = "E, dd MMM yyyy HH:mm:ss z";
+
+    private String originalDate;
+    private String amzDate;
+    private String formattedDate;
+
+    private AwsDates(String originalDate, String amzDate, String formattedDate) {
+      this.originalDate = checkNotNull(originalDate);
+      this.amzDate = checkNotNull(amzDate);
+      this.formattedDate = checkNotNull(formattedDate);
```

Comment:
It looks like this value is always computed from the `amzDate` argument which puts burden on the caller to always do the `substring(0,8)` to take the YYYMMDD part from `amzDate`. Intentional for some use-case?

Reply:
Hmm, I think having a Date instance would then burden the callers to convert the input (always a String) to a date, only to be converted back anyway.

I cleaned it up a little - you're right, there's no need to hold onto the formattedDate since it is always derived from the amzDate.